001-core/privacy.php

```diff
@@ -0,0 +1,315 @@
+	$about_group = array(
+		'group_label' => __( 'About' ),
+		'items'       => array(
+			'about-1' => array(
+				array(
+					'name'  => _x( 'Report generated for', 'email address' ),
+					'value' => $email_address,
+				),
+				array(
+					'name'  => _x( 'For site', 'website name' ),
+					'value' => get_bloginfo( 'name' ),
+				),
+				array(
+					'name'  => _x( 'At URL', 'website URL' ),
+					'value' => get_bloginfo( 'url' ),
+				),
+				array(
+					'name'  => _x( 'On', 'date/time' ),
+					'value' => current_time( 'mysql' ),
+				),
+			),
+		),
+	);
+
+	// Merge in the special about group.
+	$groups = array_merge( array( 'about' => $about_group ), $groups );
+
+	// Now, iterate over every group in $groups and have the formatter render it in HTML.
+	foreach ( (array) $groups as $group_id => $group_data ) {
+		fwrite( $file, wp_privacy_generate_personal_data_export_group_html( $group_data ) );
+	}
+
+	fwrite( $file, "</body>\n" );
+
+	// Close HTML.
+	fwrite( $file, "</html>\n" );
+	fclose( $file );
+
+	/*
+	 * Now, generate the ZIP.
+	 *
+	 * If an archive has already been generated, then remove it and reuse the
+	 * filename, to avoid breaking any URLs that may have been previously sent
+	 * via email.
+	 */
+	$error            = false;
+	$archive_url      = get_post_meta( $request_id, '_export_file_url', true );
+	$archive_pathname = get_post_meta( $request_id, '_export_file_path', true );
+
+	if ( empty( $archive_pathname ) || empty( $archive_url ) ) {
+		$archive_filename = $file_basename . '.zip';
+		$archive_pathname = $exports_dir . $archive_filename;
+		$archive_url      = $exports_url . $archive_filename;
+
+		update_post_meta( $request_id, '_export_file_url', $archive_url );
+		update_post_meta( $request_id, '_export_file_path', wp_normalize_path( $archive_pathname ) );
+	}
+
+	// Track generated time to simplify deletions.
+	// We can't currently iterate through files in the Files Service so we need a way to query exports by date.
+	update_post_meta( $request_id, '_vip_export_generated_time', time() );
+
+	// Note: core deletes the file if it exists, but we can just overwrite it when we upload.
+
+	// ZipArchive may not be available across all applications.
+	// Use it if it exists, otherwise fallback to PclZip.
+	if ( class_exists( '\ZipArchive' ) ) {
+		$zip_result = _ziparchive_create_file( $archive_pathname, $html_report_pathname );
+	} else {
+		$zip_result = _pclzip_create_file( $archive_pathname, $html_report_pathname );
+	}
+
+	if ( is_wp_error( $zip_result ) ) {
+		/* translators: %s: error message */
+		$error = sprintf( __( 'Unable to generate export file (archive) for writing: %s' ), $zip_result->get_error_message() );
+	} else {
+		/** This filter is documented in wp-admin/includes/file.php */
+		do_action( 'wp_privacy_personal_data_export_file_created', $archive_pathname, $archive_url, $html_report_pathname, $request_id );
+	}
+
+	// And remove the HTML file.
+	unlink( $html_report_pathname );
```

Comment:
Probably not a major concern right now, but I wonder if we need to consider what happens if a request times out before it gets here.

Reply:
I moved it a bit earlier in the flow which should help. The timeout, if it happens, is going to be during the zip process which needs the file, so not much we can do there.

We probably need an external cleanup routine for handling those edge cases.

Reply:
> We probably need an external cleanup routine for handling those edge cases.

Yeah, this is what I was thinking too. I think we can use `find` to list files older than X, so we could probably add a daily cron job to web containers or something.